test/cpp/interop/istio_test_echo_server.cc

```diff
@@ -0,0 +1,359 @@
+           << ")" << std::endl;
+        std::stringstream resp_ss(call->reply.message());
+        std::string line;
+        while (std::getline(resp_ss, line, '\n')) {
+          ss << "[" << call->r_id << " body] " << line << std::endl;
+        }
+        responses_->at(call->r_id) = ss.str();
+      } else
+        gpr_log(GPR_DEBUG, "RPC failed %d: %s", call->status.error_code(),
+                call->status.error_message().c_str());
+      delete call;
+    }
+  }
+
+ private:
+  struct EchoCall {
+    int r_id;  // The index of this call (5 for the 5th request sent).
+    EchoRequest request;
+    EchoResponse reply;
+    ClientContext context;
+    Status status;
+    std::unique_ptr<ClientAsyncResponseReader<EchoResponse>> response_reader;
+  };
+  std::string hostname_;
+  // TODO: the following fields are not set yet. But we may need them later.
+  //  int port_;
+  //  std::string version_;
+  //  std::string cluster_;
+  //  std::string istio_version_;
+};
+
+/*std::vector<std::unique_ptr<grpc::Server>>*/
+void RunServer(std::vector<int> ports) {
+  char base_hostname[256];
+  if (gethostname(base_hostname, 256) != 0) {
+    sprintf(base_hostname, "%s-%d", "generated", rand() % 1000);
+  }
+
+  EchoTestServiceImpl echo_test_service(base_hostname);
+  ServerBuilder builder;
+  builder.RegisterService(&echo_test_service);
+
+  for (int port : ports) {
+    std::ostringstream server_address;
+    server_address << "0.0.0.0:" << port;
+    builder.AddListeningPort(server_address.str(),
+                             grpc::InsecureServerCredentials());
+    gpr_log(GPR_DEBUG, "Server listening on %s", server_address.str().c_str());
+  }
+
+  // 3333 is the magic port that the istio testing for k8s health checks. And
+  // it only needs TCP. So also make the gRPC server to listen on 3333.
+  std::ostringstream server_address_3333;
+  server_address_3333 << "0.0.0.0:" << 3333;
+  builder.AddListeningPort(server_address_3333.str(),
+                           grpc::InsecureServerCredentials());
+  std::unique_ptr<Server> server(builder.BuildAndStart());
+  server->Wait();
+}
+
+int main(int argc, char** argv) {
+  // Preprocess argv, for two things:
+  // 1. merge duplciate flags. So "--grpc=8080 --grpc=9090" becomes
+  // "--grpc=8080,9090".
+  // 2. replace '-' to '_'. So "--istio-veriosn=123" becomes
+  // "--istio_version=123".
+  std::map<std::string, std::vector<std::string>> argv_dict;
+  for (int i = 0; i < argc; i++) {
+    std::string arg(argv[i]);
+    size_t equal = arg.find_first_of("=");
+    if (equal != std::string::npos) {
+      std::string f = arg.substr(0, equal);
+      std::string v = arg.substr(equal + 1, std::string::npos);
+      argv_dict[f].push_back(v);
+    }
+  }
+
+  std::vector<char*> new_argv_strs;
+  // Keep the command itself.
+  new_argv_strs.push_back(argv[0]);
+
+  for (std::map<std::string, std::vector<std::string>>::const_iterator it =
```

Comment:
can probably use the range based for loop instead

Reply:
Done